crosscluster/logical/txnmode: TestTxnModePauseOnEarliestConflict failed

crosscluster/logical/txnmode.TestTxnModePauseOnEarliestConflict [failed](https://mesolite.cluster.engflow.com/invocations/default/5cb51074-8b65-45dd-bd17-d54c1ea6a6b5?testReportRun=3&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvY3Jvc3NjbHVzdGVyL2xvZ2ljYWwvdHhubW9kZTp0eG5tb2RlX3Rlc3Q=) on master @ [39d3ad55242ae3760945906de6b828aabb3439f2](https://github.com/cockroachdb/cockroach/commits/39d3ad55242ae3760945906de6b828aabb3439f2):


```
=== RUN   TestTxnModePauseOnEarliestConflict
    test_log_scope.go:172: test logs captured to: outputs.zip/logTestTxnModePauseOnEarliestConflict382595098
    test_log_scope.go:82: use -show-logs to present logs inline
    test_server_shim.go:168: automatically injected a shared process virtual cluster under test; see comment at top of test_server_shim.go for details.
    txnmode_pause_on_conflict_test.go:150: condition failed to evaluate within 2m0s: expected job status paused, but got running
    panic.go:694: -- test log scope end --
test logs left over in: outputs.zip/logTestTxnModePauseOnEarliestConflict382595098
--- FAIL: TestTxnModePauseOnEarliestConflict (236.19s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=3</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/cdc
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTxnModePauseOnEarliestConflict.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-63907

crosscluster/logical/txnmode.TestTxnModePauseOnEarliestConflict [failed](https://mesolite.cluster.engflow.com/invocations/default/069d9cc5-9c52-4a4f-bd10-d2f13885c067?testReportRun=2&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvY3Jvc3NjbHVzdGVyL2xvZ2ljYWwvdHhubW9kZTp0eG5tb2RlX3Rlc3Q=) on master @ [589eff2e48e7615f14b8ba37534d13b6d556f87c](https://github.com/cockroachdb/cockroach/commits/589eff2e48e7615f14b8ba37534d13b6d556f87c):


```
=== RUN   TestTxnModePauseOnEarliestConflict
    test_log_scope.go:172: test logs captured to: outputs.zip/logTestTxnModePauseOnEarliestConflict3827271619
    test_log_scope.go:82: use -show-logs to present logs inline
    test_server_shim.go:168: automatically injected a shared process virtual cluster under test; see comment at top of test_server_shim.go for details.
    txnmode_pause_on_conflict_test.go:150: condition failed to evaluate within 2m0s: expected job status paused, but got running
    panic.go:694: -- test log scope end --
test logs left over in: outputs.zip/logTestTxnModePauseOnEarliestConflict3827271619
--- FAIL: TestTxnModePauseOnEarliestConflict (216.71s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=2</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTxnModePauseOnEarliestConflict.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/crosscluster/logical/resume_txn.go
+++ b/pkg/crosscluster/logical/resume_txn.go
@@ -26,16 +26,17 @@ import (
 
 // resumeTransactionalLdr runs the transactional LDR ingestion loop, converging
 // on the earliest unappliable transaction before pausing the job.
+//
 // The control loop has three outcomes per attempt:
 //  1. The coordinator returns a txnapply.ReplicationError at timestamp T:
-//     capture it as the candidate pause cause, shrink endTime to T, and retry.
-//     The next attempt's mergefeed will skip events at and after T, so a
-//     conflicting txn at T can never re-trigger the same error; instead, an
+//     capture it as the candidate pause cause, shrink endTime to T.Prev(), and
+//     retry. The next attempt's mergefeed will skip events at and after T, so
+//     a conflicting txn at T can never re-trigger the same error; instead, an
 //     earlier conflict (if any) will surface.
 //  2. The coordinator returns ErrEndTimeReached: every txn with timestamp
-//     strictly less than endTime has been applied without surfacing a new
-//     conflict, so the captured ReplicationError identifies the earliest
-//     conflicting transaction. Pause the job with that error.
+//     <= endTime has been applied without surfacing a new conflict, so the
+//     captured ReplicationError identifies the earliest conflicting
+//     transaction. Pause the job with that error.
 //  3. Any other error: propagate it; resumeWithRetries decides whether to
 //     retry or surface the failure.
 func (r *logicalReplicationResumer) resumeTransactionalLdr(
@@ -47,7 +48,7 @@ func (r *logicalReplicationResumer) resumeTransactionalLdr(
 		err := r.runTxnCoordinator(ctx, jobExecCtx, endTime)
 		switch {
 		case errors.As(err, &capturedErr):
-			endTime = capturedErr.Timestamp
+			endTime = capturedErr.Timestamp.Prev()
 			return err
 		case errors.Is(err, txnmode.ErrEndTimeReached):
 			if capturedErr == nil {
@@ -68,6 +69,18 @@ func (r *logicalReplicationResumer) resumeTransactionalLdr(
 func (r *logicalReplicationResumer) runTxnCoordinator(
 	ctx context.Context, jobExecCtx sql.JobExecContext, endTime hlc.Timestamp,
 ) error {
+	// Skip setup entirely if the job has already replicated past endTime. This
+	// happens when the convergence loop has shrunk endTime below the persisted
+	// frontier and there is no work left in the [replicatedTime, endTime]
+	// window.
+	replicatedTime, err := replicatedTimeFromJob(ctx, jobExecCtx.ExecCfg().InternalDB, r.job)
+	if err != nil {
+		return err
+	}
+	if endTime.LessEq(replicatedTime) {
+		return txnmode.ErrEndTimeReached
+	}
+
 	client, err := r.getActiveClient(ctx, jobExecCtx.ExecCfg().InternalDB)
 	if err != nil {
 		return err
@@ -87,11 +100,6 @@ func (r *logicalReplicationResumer) runTxnCoordinator(
 	// TODO(jeffswenson): checkpoint partition URIs via
 	// r.checkpointPartitionURIs once plan generation is added.
 
-	// Build the DistSQL physical plan before starting concurrent work.
-	replicatedTime, err := replicatedTimeFromJob(ctx, jobExecCtx.ExecCfg().InternalDB, r.job)
-	if err != nil {
-		return err
-	}
 	flowPlan, planCtx, applierInstanceIDs, err :=
 		txnmode.PlanTxnReplication(ctx, r.job, jobExecCtx, sourcePlan, replicatedTime, endTime)
 	if err != nil {

--- a/pkg/crosscluster/logical/txnapply/BUILD.bazel
+++ b/pkg/crosscluster/logical/txnapply/BUILD.bazel
@@ -27,6 +27,7 @@ go_library(
         "//pkg/util/syncutil",
         "@com_github_cockroachdb_errors//:errors",
         "@com_github_cockroachdb_redact//:redact",
+        "@com_github_gogo_protobuf//proto",
     ],
 )
 

--- a/pkg/crosscluster/logical/txnapply/txn_applier.go
+++ b/pkg/crosscluster/logical/txnapply/txn_applier.go
@@ -22,7 +22,14 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
 	"github.com/cockroachdb/errors"
 	"github.com/cockroachdb/redact"
+	"github.com/gogo/protobuf/proto"
 )
+
+func init() {
+	pKey := errors.GetTypeKey((*ReplicationError)(nil))
+	errors.RegisterWrapperEncoder(pKey, encodeReplicationError)
+	errors.RegisterWrapperDecoder(pKey, decodeReplicationError)
+}
 
 var leaseReleaseInterval = settings.RegisterDurationSetting(
 	settings.ApplicationLevel,
@@ -53,6 +60,25 @@ func (e *ReplicationError) SafeFormat(p redact.SafePrinter, _ rune) {
 
 // Unwrap implements the errors.Wrapper interface.
 func (e *ReplicationError) Unwrap() error { return e.Err }
+
+func encodeReplicationError(
+	_ context.Context, err error,
+) (msgPrefix string, safe []string, details proto.Message) {
+	var e *ReplicationError
+	errors.As(err, &e)
+	return "", nil, &e.Timestamp
+}
+
+func decodeReplicationError(
+	_ context.Context, cause error, _ string, _ []string, payload proto.Message,
+) error {
+	ts, ok := payload.(*hlc.Timestamp)
+	if !ok {
+		// Payload type changed unexpectedly; fall back to opaque decode.
+		return nil
+	}
+	return &ReplicationError{Timestamp: *ts, Err: cause}
+}
 
 type appliedTransaction struct {
 	ldrdecoder.Transaction

--- a/pkg/crosscluster/logical/txnfeed/merge_feed.go
+++ b/pkg/crosscluster/logical/txnfeed/merge_feed.go
@@ -30,8 +30,8 @@ import (
 // encountered). This guarantees that a transaction is never split across
 // batches.
 //
-// The merge feed only emits events with timestamp strictly less than endTime.
-// After encountering the first event at or beyond endTime, it flushes any
+// The merge feed only emits events with timestamp less than or equal to endTime.
+// After encountering the first event beyond endTime, it flushes any
 // buffered KVs and emits a synthetic checkpoint at endTime.Prev(). For all
 // remaining events, it continues normal heap processing but only ever emits
 // the synthetic checkpoint (deduplicated by maybeEmitCheckpoint) until every
@@ -113,7 +113,7 @@ func (m *MergeFeed) mergeLoop(ctx context.Context) error {
 	for h.len() != 0 {
 		entry := h.pop()
 
-		if m.endTime.LessEq(entry.peekTS) {
+		if m.endTime.Less(entry.peekTS) {
 			if err := m.processPastEndTime(ctx, entry, h); err != nil {
 				return err
 			}
@@ -204,14 +204,14 @@ func (m *MergeFeed) initHeap(ctx context.Context) (*mergeHeap, error) {
 	return h, nil
 }
 
-// processPastEndTime drains an entry whose timestamp is at or beyond endTime.
-// It flushes any buffered KVs, emits a synthetic checkpoint at endTime.Prev(),
-// and advances the entry without processing it.
+// processPastEndTime drains an entry whose timestamp is past endTime.
+// It flushes any buffered KVs, emits a synthetic checkpoint at endTime, and
+// advances the entry without processing it.
 func (m *MergeFeed) processPastEndTime(ctx context.Context, entry *mergeEntry, h *mergeHeap) error {
 	if err := m.flushScratch(ctx); err != nil {
 		return err
 	}
-	if err := m.emitCheckpoint(ctx, m.endTime.Prev()); err != nil {
+	if err := m.emitCheckpoint(ctx, m.endTime); err != nil {
 		return err
 	}
 	var closed bool

--- a/pkg/crosscluster/logical/txnfeed/merge_feed_test.go
+++ b/pkg/crosscluster/logical/txnfeed/merge_feed_test.go
@@ -291,6 +291,7 @@ func TestMergeFeedEndTime(t *testing.T) {
 		sub2 := makeTestSubscription([]any{
 			kvEvent{key: "b3", time: 3},
 			kvEvent{key: "b6", time: 6},
+			kvEvent{key: "b7", time: 7},
 			kvEvent{key: "b8", time: 8},
 			checkpoint{start: "a", end: "b", time: 8},
 			closeEvent{},
@@ -322,18 +323,18 @@ func TestMergeFeedEndTime(t *testing.T) {
 				}
 			case crosscluster.CheckpointEvent:
 				cp := ev.GetCheckpoint().ResolvedSpans[0].Timestamp
-				require.True(t, cp.Less(endTime),
-					"checkpoints at or beyond endTime must not be emitted")
-				if cp.Equal(endTime.Prev()) {
+				require.True(t, cp.LessEq(endTime),
+					"checkpoints past endTime must not be emitted")
+				if cp.Equal(endTime) {
 					sawSyntheticCP = true
 				}
 			}
 		}
 
-		// Only KVs with timestamp strictly less than endTime should be emitted.
+		// Only KVs with timestamp <= endTime should be emitted.
 		var expectedKVTimes []int
-		for _, ts := range []int{1, 3, 5, 6, 8, 10} {
-			if (hlc.Timestamp{WallTime: int64(ts)}).Less(endTime) {
+		for _, ts := range []int{1, 3, 5, 6, 7, 8, 10} {
+			if (hlc.Timestamp{WallTime: int64(ts)}).LessEq(endTime) {
 				expectedKVTimes = append(expectedKVTimes, ts)
 			}
 		}

--- a/pkg/crosscluster/logical/txnmode/BUILD.bazel
+++ b/pkg/crosscluster/logical/txnmode/BUILD.bazel
@@ -97,6 +97,7 @@ go_test(
         "//pkg/sql/isql",
         "//pkg/sql/physicalplan",
         "//pkg/sql/sem/tree",
+        "//pkg/testutils",
         "//pkg/testutils/jobutils",
         "//pkg/testutils/serverutils",
         "//pkg/testutils/skip",

--- a/pkg/crosscluster/logical/txnmode/alter_logical_replication_test.go
+++ b/pkg/crosscluster/logical/txnmode/alter_logical_replication_test.go
@@ -29,9 +29,9 @@ func TestAlterLDRSkipAdvancesReplicatedTime(t *testing.T) {
 	defer log.Scope(t).Close(t)
 	ctx := context.Background()
 
-	srv, sourceDB, destDB := setupTxnModeTest(t)
-	defer srv.Stopper().Stop(ctx)
-	jobID := setupConflictingLDR(t, srv, sourceDB, destDB)
+	cluster, sourceDB, destDB := setupTxnModeTest(t, 1 /* numNodes */)
+	defer cluster.Stopper().Stop(ctx)
+	jobID := setupConflictingLDR(t, cluster, sourceDB, destDB)
 	jobutils.WaitForJobToPause(t, destDB, jobID)
 
 	replicatedBefore, err := ldrtestutils.GetReplicatedTime(t, destDB, jobID)
@@ -47,7 +47,7 @@ func TestAlterLDRSkipAdvancesReplicatedTime(t *testing.T) {
 	jobutils.WaitForJobToRun(t, destDB, jobID)
 
 	sourceDB.Exec(t, "INSERT INTO tab VALUES (2, 'after-skip')")
-	ldrtestutils.WaitUntilReplicatedTime(t, srv.Clock().Now(), destDB, jobID)
+	ldrtestutils.WaitUntilReplicatedTime(t, cluster.Server(0).Clock().Now(), destDB, jobID)
 
 	destDB.CheckQueryResults(t,
 		"SELECT pk, val FROM tab ORDER BY pk",
@@ -61,14 +61,14 @@ func TestAlterLDRSkipRejectsInvalidState(t *testing.T) {
 	defer log.Scope(t).Close(t)
 	ctx := context.Background()
 
-	srv, sourceDB, destDB := setupTxnModeTest(t)
-	defer srv.Stopper().Stop(ctx)
+	cluster, sourceDB, destDB := setupTxnModeTest(t, 1 /* numNodes */)
+	defer cluster.Stopper().Stop(ctx)
 
 	for _, db := range []*sqlutils.SQLRunner{sourceDB, destDB} {
 		db.Exec(t, "CREATE TABLE tab (pk INT PRIMARY KEY, val STRING NOT NULL)")
 	}
 
-	s := srv.ApplicationLayer()
+	s := cluster.Server(0).ApplicationLayer()
 	sourceURL := replicationtestutils.GetExternalConnectionURI(t, s, s, serverutils.DBName("source_db"))
 
 	t.Run("skip on running job", func(t *testing.T) {

--- a/pkg/crosscluster/logical/txnmode/ldr_applier_processor.go
+++ b/pkg/crosscluster/logical/txnmode/ldr_applier_processor.go
@@ -25,6 +25,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/admission/admissionpb"
 	"github.com/cockroachdb/cockroach/pkg/util/ctxgroup"
 	"github.com/cockroachdb/cockroach/pkg/util/hlc"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/cockroach/pkg/util/protoutil"
 	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
 	"github.com/cockroachdb/errors"
@@ -65,10 +66,12 @@ type ldrApplierProcessor struct {
 	// grp manages the lifecycle of the input reader, applier, and
 	// backchannel forwarder goroutines. cancelGrp cancels the context
 	// used by grp, allowing close() to unblock goroutines waiting on
-	// channel operations.
+	// channel operations. errCh is where errors from goroutines are
+	// sent.
 	grp       ctxgroup.Group
 	cancelGrp context.CancelFunc
 	grpCtx    context.Context
+	errCh     chan error
 }
 
 // Start implements execinfra.RowSource.
@@ -89,22 +92,40 @@ func (p *ldrApplierProcessor) Start(ctx context.Context) {
 	// Read coordinator input rows, deserialize, and feed to the Applier.
 	p.grp.GoCtx(func(ctx context.Context) error {
 		defer close(p.applierEvents)
-		return p.runInputReader(ctx)
+		p.sendError(p.runInputReader(ctx))
+		return nil
 	})
 
 	// Run the Applier's internal pipeline.
 	p.grp.GoCtx(func(ctx context.Context) error {
-		// Closing the applier here ensures the frontier channel closes, allowing
-		// Next() to proceed if the applier errrors.
 		defer p.applier.Close(ctx)
-		return p.applier.Run(ctx, p.applierEvents)
+		p.sendError(p.applier.Run(ctx, p.applierEvents))
+		return nil
 	})
 
 	// Forward loopback updates from the dep resolver to the Receive()
 	// channel or to the output.
 	p.grp.GoCtx(func(ctx context.Context) error {
-		return p.depResolver.RunBackchannelForwarder(ctx, p.loopbackChs.updateCh)
+		p.sendError(p.depResolver.RunBackchannelForwarder(ctx, p.loopbackChs.updateCh))
+		return nil
 	})
+}
+
+// sendError returns any non-cancellation error from the applier to
+// the errCh. We opt against plumbing the error to the trailing metadata
+// as it is cleaner to tear down the flow by cancelling the context on error
+// than to try and propagate the error to other appliers; trailing metadata
+// is not propagated until draining is complete, but we need the error to
+// signal draining for other appliers.
+func (p *ldrApplierProcessor) sendError(err error) {
+	if err == nil || errors.Is(err, context.Canceled) {
+		return
+	}
+	select {
+	case p.errCh <- err:
+	default:
+		log.Dev.VInfof(p.grpCtx, 2, "dropping additional error: %s", err)
+	}
 }
 
 // setup initializes all components required for the applier pipeline:
@@ -166,6 +187,7 @@ func (p *ldrApplierProcessor) setup(ctx context.Context) error {
 	}
 
 	p.applierEvents = make(chan txnapply.ApplierEvent)
+	p.errCh = make(chan error, 1)
 
 	p.loopbackChs = ldrLoopback.lookupOrCreate(p.FlowCtx, p.spec.ApplierID)
 	return nil
@@ -176,6 +198,12 @@ func (p *ldrApplierProcessor) setup(ctx context.Context) error {
 func (p *ldrApplierProcessor) Next() (rowenc.EncDatumRow, *execinfrapb.ProducerMetadata) {
 	for p.State == execinfra.StateRunning {
 		select {
+		case <-p.Ctx().Done():
+			p.MoveToDraining(nil)
+
+		case err := <-p.errCh:
+			p.MoveToDraining(err)
+
 		case <-p.depResolver.OutCh():
 			ev, ok := p.depResolver.PopOutEvent()
 			if !ok {
@@ -190,9 +218,9 @@ func (p *ldrApplierProcessor) Next() (rowenc.EncDatumRow, *execinfrapb.ProducerM
 
 		case frontier, ok := <-p.applier.Frontier():
 			if !ok {
-				p.MoveToDraining(nil)
 				break
 			}
+
 			meta, err := p.encodeFrontierMeta(frontier)
 			if err != nil {
 				p.MoveToDraining(err)
@@ -209,12 +237,10 @@ func (p *ldrApplierProcessor) ConsumerClosed() {
 }
 
 // close cancels background goroutines, waits for them to exit, and cleans up
-// resources. It returns any non-cancellation error from the ctxgroup as
-// trailing metadata so the TrailingMetaCallback can plumb it back to the
-// distsql flow consumer.
-func (p *ldrApplierProcessor) close() []execinfrapb.ProducerMetadata {
+// resources.
+func (p *ldrApplierProcessor) close() {
 	if p.Closed {
-		return nil
+		return
 	}
 
 	// Cancel the goroutine context to unblock any goroutines waiting on
@@ -222,13 +248,13 @@ func (p *ldrApplierProcessor) close() []execinfrapb.ProducerMetadata {
 	if p.cancelGrp != nil {
 		p.cancelGrp()
 	}
-	var meta []execinfrapb.ProducerMetadata
 	if err := p.grp.Wait(); err != nil && !errors.Is(err, context.Canceled) {
-		meta = append(meta, execinfrapb.ProducerMetadata{Err: err})
+		// We should be propagating errors through the errCh, so seeing
+		// one here would be unexpected.
+		log.Dev.Warningf(p.grpCtx, "dropping additional error: %s", err)
 	}
 
 	p.InternalClose()
-	return meta
 }
 
 // runInputReader reads input rows from the coordinator, deserializes them
@@ -361,7 +387,8 @@ func init() {
 			execinfra.ProcStateOpts{
 				InputsToDrain: []execinfra.RowSource{input},
 				TrailingMetaCallback: func() []execinfrapb.ProducerMetadata {
-					return proc.close()
+					proc.close()
+					return nil
 				},
 			},
 		)

--- a/pkg/crosscluster/logical/txnmode/settings.go
+++ b/pkg/crosscluster/logical/txnmode/settings.go
@@ -13,7 +13,7 @@ import (
 )
 
 // ErrEndTimeReached signals that the coordinator has processed all transactions
-// with timestamp strictly less than endTime.
+// with timestamp less than or equal to endTime.
 var ErrEndTimeReached = errors.New("LDR endTime reached")
 
 var txnNumWriters = settings.RegisterIntSetting(

--- a/pkg/crosscluster/logical/txnmode/txnmode_dist.go
+++ b/pkg/crosscluster/logical/txnmode/txnmode_dist.go
@@ -392,6 +392,11 @@ func newCheckpointHandler(
 func (ch *checkpointHandler) handleMeta(
 	ctx context.Context, meta *execinfrapb.ProducerMetadata,
 ) error {
+	if meta.Err != nil {
+		ch.cancelFlow()
+		return nil
+	}
+
 	if meta.BulkProcessorProgress == nil ||
 		len(meta.BulkProcessorProgress.ProgressMessage) == 0 {
 		return nil

--- a/pkg/crosscluster/logical/txnmode/txnmode_dist.go
+++ b/pkg/crosscluster/logical/txnmode/txnmode_dist.go
@@ -417,10 +417,11 @@ func (ch *checkpointHandler) handleMeta(
 		return nil
 	}
 
-	reachedEndTime := !replicatedTime.Less(ch.endTime.Prev())
-	if replicatedTime.LessEq(ch.replicatedTime) && !reachedEndTime {
+	if replicatedTime.LessEq(ch.replicatedTime) {
 		return nil
 	}
+
+	reachedEndTime := ch.endTime.LessEq(replicatedTime)
 	updateFreq := ldrsettings.JobCheckpointFrequency.Get(ch.sv)
 	if !reachedEndTime && (updateFreq == 0 || timeutil.Since(ch.lastPersistenceTime) < updateFreq) {
 		return nil

--- a/pkg/crosscluster/logical/txnmode/txnmode_dist_test.go
+++ b/pkg/crosscluster/logical/txnmode/txnmode_dist_test.go
@@ -458,12 +458,12 @@ func TestCheckpointHandler(t *testing.T) {
 		require.Len(t, updates, 1)
 		require.Equal(t, mkts(15), updates[0])
 
-		// Applier 1 advances: min frontier is ts(18), still below endTime.Prev().
+		// Applier 1 advances: min frontier is ts(18), still below endTime.
 		require.NoError(t, ch.handleMeta(ctx, makeFrontierMeta(1, mkts(25))))
 		updates = drainChannel(frontierCh)
 		require.Equal(t, mkts(18), updates[0])
 
-		// Applier 2 advances: min becomes ts(25) which is >= endTime.Prev().
+		// Applier 2 advances: min becomes ts(25) which is > endTime.
 		err := ch.handleMeta(ctx, makeFrontierMeta(2, mkts(30)))
 		require.ErrorIs(t, err, ErrEndTimeReached)
 		require.True(t, flowCanceled)

--- a/pkg/crosscluster/logical/txnmode/txnmode_pause_on_conflict_test.go
+++ b/pkg/crosscluster/logical/txnmode/txnmode_pause_on_conflict_test.go
@@ -15,35 +15,38 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/crosscluster/replicationtestutils"
 	"github.com/cockroachdb/cockroach/pkg/jobs"
 	"github.com/cockroachdb/cockroach/pkg/jobs/jobspb"
+	"github.com/cockroachdb/cockroach/pkg/testutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/jobutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/skip"
 	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
+	"github.com/cockroachdb/cockroach/pkg/testutils/testcluster"
 	"github.com/cockroachdb/cockroach/pkg/util/hlc"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/stretchr/testify/require"
 )
 
-// setupTxnModeTest starts a server with source and destination databases
+// setupTxnModeTest starts a cluster with source and destination databases
 // configured for low-latency replication. It is the caller's responsibility
-// to stop the returned server.
+// to stop the returned cluster.
 func setupTxnModeTest(
-	t *testing.T,
-) (serverutils.TestServerInterface, *sqlutils.SQLRunner, *sqlutils.SQLRunner) {
+	t *testing.T, numNodes int,
+) (*testcluster.TestCluster, *sqlutils.SQLRunner, *sqlutils.SQLRunner) {
 	t.Helper()
-
-	srv, conn, _ := serverutils.StartServer(t, base.TestServerArgs{
-		DefaultTestTenant: base.TestDoesNotWorkWithExternalProcessMode(134857),
-		Knobs: base.TestingKnobs{
-			JobsTestingKnobs: jobs.NewTestingKnobsWithShortIntervals(),
+	cluster := testcluster.StartTestCluster(t, numNodes, base.TestClusterArgs{
+		ServerArgs: base.TestServerArgs{
+			DefaultTestTenant: base.TestDoesNotWorkWithExternalProcessMode(134857),
+			Knobs: base.TestingKnobs{
+				JobsTestingKnobs: jobs.NewTestingKnobsWithShortIntervals(),
+			},
 		},
 	})
 
-	s := srv.ApplicationLayer()
-	runner := sqlutils.MakeSQLRunner(conn)
+	s := cluster.Server(0).ApplicationLayer()
+	runner := sqlutils.MakeSQLRunner(cluster.Conns[0])
 
-	sysRunner := sqlutils.MakeSQLRunner(srv.SystemLayer().SQLConn(t))
+	sysRunner := sqlutils.MakeSQLRunner(cluster.SystemLayer(0).SQLConn(t))
 	ldrtestutils.ApplyLowLatencyReplicationSettings(t, sysRunner, runner)
 
 	runner.Exec(t, "CREATE DATABASE source_db")
@@ -52,7 +55,7 @@ func setupTxnModeTest(
 	sourceDB := sqlutils.MakeSQLRunner(s.SQLConn(t, serverutils.DBName("source_db")))
 	destDB := sqlutils.MakeSQLRunner(s.SQLConn(t, serverutils.DBName("dest_db")))
 
-	return srv, sourceDB, destDB
+	return cluster, sourceDB, destDB
 }
 
 // setupConflictingLDR creates a table with a unique index on both source and
@@ -61,7 +64,7 @@ func setupTxnModeTest(
 // The caller is responsible for waiting on the job state.
 func setupConflictingLDR(
 	t *testing.T,
-	srv serverutils.TestServerInterface,
+	cluster *testcluster.TestCluster,
 	sourceDB *sqlutils.SQLRunner,
 	destDB *sqlutils.SQLRunner,
 ) jobspb.JobID {
@@ -74,8 +77,8 @@ func setupConflictingLDR(
 
 	destDB.Exec(t, "INSERT INTO tab VALUES (100, 'collide')")
 
-	sourceURL := replicationtestutils.GetExternalConnectionURI(
-		t, srv.ApplicationLayer(), srv.ApplicationLayer(), serverutils.DBName("source_db"))
+	s := cluster.Server(0).ApplicationLayer()
+	sourceURL := replicationtestutils.GetExternalConnectionURI(t, s, s, serverutils.DBName("source_db"))
 
 	var jobID jobspb.JobID
 	destDB.QueryRow(t,
@@ -94,32 +97,39 @@ func TestTxnModePauseOnConflict(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	skip.UnderDeadlock(t)
 	defer log.Scope(t).Close(t)
-	ctx := context.Background()
 
-	srv, sourceDB, destDB := setupTxnModeTest(t)
-	defer srv.Stopper().Stop(ctx)
-	jobID := setupConflictingLDR(t, srv, sourceDB, destDB)
-	jobutils.WaitForJobToPause(t, destDB, jobID)
+	testutils.RunValues(t, "nodes", []int{1, 3}, func(t *testing.T, numNodes int) {
+		if numNodes > 1 {
+			// Multi node test clusters can timeout under stress+race.
+			skip.UnderDuress(t)
+		}
+		ctx := context.Background()
 
-	var runningStatus string
-	destDB.QueryRow(t, "SELECT running_status FROM [SHOW JOBS] WHERE job_id = $1", jobID).Scan(&runningStatus)
-	require.Contains(t, runningStatus, "replication error")
-	require.Contains(t, runningStatus, "duplicate key value violates unique constraint")
+		cluster, sourceDB, destDB := setupTxnModeTest(t, numNodes)
+		defer cluster.Stopper().Stop(ctx)
+		jobID := setupConflictingLDR(t, cluster, sourceDB, destDB)
+		jobutils.WaitForJobToPause(t, destDB, jobID)
 
-	// The conflicting source row at pk=1 must not have been applied.
-	destDB.CheckQueryResults(t, "SELECT pk, val FROM tab ORDER BY pk", [][]string{
-		{"100", "collide"},
+		var runningStatus string
+		destDB.QueryRow(t, "SELECT running_status FROM [SHOW JOBS] WHERE job_id = $1", jobID).Scan(&runningStatus)
+		require.Contains(t, runningStatus, "replication error")
+		require.Contains(t, runningStatus, "duplicate key value violates unique constraint")
+
+		// The conflicting source row at pk=1 must not have been applied.
+		destDB.CheckQueryResults(t, "SELECT pk, val FROM tab ORDER BY pk", [][]string{
+			{"100", "collide"},
+		})
+
+		// Check that the replicated time equals the MVCC timestamp of the
+		// conflicting source insert minus one logical tick.
+		var conflictMVCCDec apd.Decimal
+		sourceDB.QueryRow(t, "SELECT crdb_internal_mvcc_timestamp FROM tab WHERE pk = 1").Scan(&conflictMVCCDec)
+		conflictMVCC, err := hlc.DecimalToHLC(&conflictMVCCDec)
+		require.NoError(t, err)
+		replicatedTime, err := ldrtestutils.GetReplicatedTime(t, destDB, jobID)
+		require.NoError(t, err)
+		require.Equal(t, conflictMVCC.Prev(), replicatedTime)
 	})
-
-	// Check that the replicated time equals the MVCC timestamp of the
-	// conflicting source insert minus one logical tick.
-	var conflictMVCCDec apd.Decimal
-	sourceDB.QueryRow(t, "SELECT crdb_internal_mvcc_timestamp FROM tab WHERE pk = 1").Scan(&conflictMVCCDec)
-	conflictMVCC, err := hlc.DecimalToHLC(&conflictMVCCDec)
-	require.NoError(t, err)
-	replicatedTime, err := ldrtestutils.GetReplicatedTime(t, destDB, jobID)
-	require.NoError(t, err)
-	require.Equal(t, conflictMVCC.Prev(), replicatedTime)
 }
 
 // TestTxnModeResumeAfterFixingConflict verifies that when a transactional LDR
@@ -132,9 +142,9 @@ func TestTxnModeResumeAfterFixingConflict(t *testing.T) {
 	defer log.Scope(t).Close(t)
 	ctx := context.Background()
 
-	srv, sourceDB, destDB := setupTxnModeTest(t)
-	defer srv.Stopper().Stop(ctx)
-	jobID := setupConflictingLDR(t, srv, sourceDB, destDB)
+	cluster, sourceDB, destDB := setupTxnModeTest(t, 1 /* numNodes */)
+	defer cluster.Stopper().Stop(ctx)
+	jobID := setupConflictingLDR(t, cluster, sourceDB, destDB)
 	jobutils.WaitForJobToPause(t, destDB, jobID)
 
 	// Remove the conflicting row on the destination so the transaction
@@ -144,7 +154,7 @@ func TestTxnModeResumeAfterFixingConflict(t *testing.T) {
 	destDB.Exec(t, "RESUME JOB $1", jobID)
 	jobutils.WaitForJobToRun(t, destDB, jobID)
 
-	now := srv.Clock().Now()
+	now := cluster.Server(0).Clock().Now()
 	ldrtestutils.WaitUntilReplicatedTime(t, now, destDB, jobID)
 
 	destDB.CheckQueryResults(t,
@@ -162,9 +172,9 @@ func TestTxnModeResumePausesAgainOnUnresolvedConflict(t *testing.T) {
 	defer log.Scope(t).Close(t)
 	ctx := context.Background()
 
-	srv, sourceDB, destDB := setupTxnModeTest(t)
-	defer srv.Stopper().Stop(ctx)
-	jobID := setupConflictingLDR(t, srv, sourceDB, destDB)
+	cluster, sourceDB, destDB := setupTxnModeTest(t, 1 /* numNodes */)
+	defer cluster.Stopper().Stop(ctx)
+	jobID := setupConflictingLDR(t, cluster, sourceDB, destDB)
 	jobutils.WaitForJobToPause(t, destDB, jobID)
 
 	progressFirst := jobutils.GetJobProgress(t, destDB, jobID)
@@ -188,8 +198,8 @@ func TestTxnModePauseOnEarliestConflict(t *testing.T) {
 	defer log.Scope(t).Close(t)
 	ctx := context.Background()
 
-	srv, sourceDB, destDB := setupTxnModeTest(t)
-	defer srv.Stopper().Stop(ctx)
+	cluster, sourceDB, destDB := setupTxnModeTest(t, 1 /* numNodes */)
+	defer cluster.Stopper().Stop(ctx)
 
 	for _, db := range []*sqlutils.SQLRunner{sourceDB, destDB} {
 		db.Exec(t, "CREATE TABLE tab (pk INT PRIMARY KEY, val STRING NOT NULL, extra STRING NOT NULL)")
@@ -199,7 +209,7 @@ func TestTxnModePauseOnEarliestConflict(t *testing.T) {
 
 	destDB.Exec(t, "INSERT INTO tab VALUES (100, 'first-collide', 'pre-1'), (101, 'pre-2', 'second-collide')")
 
-	s := srv.ApplicationLayer()
+	s := cluster.Server(0).ApplicationLayer()
 	sourceURL := replicationtestutils.GetExternalConnectionURI(t, s, s, serverutils.DBName("source_db"))
 
 	var jobID jobspb.JobID

--- a/pkg/sql/execinfrapb/processors_txn_ldr.proto
+++ b/pkg/sql/execinfrapb/processors_txn_ldr.proto
@@ -63,7 +63,7 @@ message TxnLDRCoordinatorSpec {
 
   optional LDRSchema schema = 6 [(gogoproto.nullable) = false];
 
-  // The coordinator processes only events with timestamp strictly less than
+  // The coordinator processes only events with timestamp less than or equal to
   // end_time and returns ErrEndTimeReached once the merge feed reaches this
   // cutoff. Used by the pause-on-conflict convergence loop.
   optional util.hlc.Timestamp end_time = 7 [(gogoproto.nullable) = false];

--- a/pkg/testutils/lint/lint_test.go
+++ b/pkg/testutils/lint/lint_test.go
@@ -1639,6 +1639,7 @@ func TestLint(t *testing.T) {
 			":!*.pb.go",
 			":!*.pb.gw.go",
 			":!ccl/changefeedccl/changefeedbase/errors.go",
+			":!crosscluster/logical/txnapply/txn_applier.go",
 			":!kv/kvclient/kvcoord/lock_spans_over_budget_error.go",
 			":!kv/kvserver/kvserverbase/bulk_adder.go",
 			":!spanconfig/errors.go",

